lib/compact.js

```diff
@@ -220,7 +221,15 @@ api.compact = async ({
         // use keyword alias and add value
         const alias = api.compactIri(
           {activeCtx, iri: expandedProperty, relativeTo: {vocab: true}});
-        const isArray = _isArray(compactedValue) && expandedValue.length === 0;
+        const container = _getContextValue(
+          activeCtx, alias, '@container') || [];
+
+        // treat as array for @type if @container includes @set
+        const typeAsSet = isType &&
+          container.includes('@set') &&
+          _processingMode(activeCtx, 1.1);
+        const isArray =
+          (_isArray(compactedValue) && expandedValue.length === 0) || typeAsSet;
```

Comment:
Potential micro optimization?  Check `typeAsSet` first since it's already calculated and maybe can short circuit the other tests.  Double check this logic please.
```suggestion
          typeAsSet || (_isArray(compactedValue) && expandedValue.length === 0);
```